File: 1930-unique-length-3-palindromic-subsequences/1930-unique-length-3-palindromic-subsequences.py
class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        res = 0
        n = len(s)
        first = [-1] * 26
        last = [-1] * 26
        for i in range(n):
            c = ord(s[i]) - ord('a')
            if first[c] == -1:
                first[c] = i
            last[c] = i
        
        for i in range(26):
            if first[i] == -1 or last[i] == -1 or first[i] == last[i]: continue
            tmp = set()
            for j in range(first[i] + 1, last[i]):
                tmp.add(s[j])

            res += len(tmp)

        return res

File: 1930-unique-length-3-palindromic-subsequences/1930-unique-length-3-palindromic-subsequences.rs
use std::collections::HashSet;

impl Solution {
    pub fn count_palindromic_subsequence(s: String) -> i32 {
        let n: usize = s.len();
        let mut first = vec![n; 26];
        let mut last = vec![n; 26];
        let mut res = 0;
        let s = s.as_bytes();

        for (i, &c) in s.iter().enumerate() {
            let idx = (c - b'a') as usize;
            if first[idx] == n {
                first[idx] = i;
            }
            last[idx] = i;
        }

        for i in 0..26 {
            if first[i] == n {
                continue;
            }
            let mut tmp = HashSet::new();
            for j in first[i]+1..last[i] {
                tmp.insert(s[j]);
            }
            res += tmp.len();
        }

        res as i32
    }
}